Add benchmarks for Frame writes

cesarbs opened the pull request:
@halter73 @pakrym

benaadams commented on test/Microsoft.AspNetCore.Server.Kestrel.Performance/project.json:
```diff
@@ -2,7 +2,8 @@
   "version": "1.0.0-*",
   "dependencies": {
     "BenchmarkDotNet": "0.10.0",
```
Bump to 10.1 and netcoreapp1.1? as per https://github.com/aspnet/KestrelHttpServer/pull/1222/commits/8301721b80d57828fb106699685502c1d0ff5f58

Gives inbuilt accurate memory profiling

halter73 commented on test/Microsoft.AspNetCore.Server.Kestrel.Performance/Writing.cs:
```diff
@@ -0,0 +1,73 @@
+            _frame.InitializeHeaders();
+        }
+
+        [Benchmark]
```
I think we should have one variation for a set Content-Length and another for chunked responses.

benaadams commented on test/Microsoft.AspNetCore.Server.Kestrel.Performance/Writing.cs:
```diff
@@ -0,0 +1,125 @@
+        [Benchmark(OperationsPerInvoke = InnerLoopCount)]
+        public void Write()
+        {
+            _frameWrite.Write(new ArraySegment<byte>(_writeData));
```
Should these functions have a for loop in them?

e.g.
```csharp
for (var i = 0; i < InnerLoopCount; i++)
{
    _frameWrite.Write(new ArraySegment<byte>(_writeData));
}
```
The attribute `OperationsPerInvoke` says how much to divide the total time by

cesarbs replied:
I guess I misunderstood how the attribute works. Will change.

benaadams replied:
It tells it how many times you loop; so you can measure with less function call overhead by adding a loop

cesarbs commented:
Made things simpler - doing 1-byte writes and setting `Content-Length` to a large enough value that it doesn't throw because of too many bytes written.

I'm having an issue though: the first benchmark is always bombed:

```
                   Method |        Mean |    StdErr |    StdDev |  Gen 0 | Allocated |
------------------------- |------------ |---------- |---------- |------- |---------- |
                    Write |          NA |        NA |        NA |    N/A |       N/A |
```

I'm seeing this error when it gets run:

```
// Validating benchmarks:
// **************************
// Benchmark: Writing.Write: DefaultJob
// *** Generate ***
// Result = Success
// BinariesDirectoryPath = D:\src\aspnet\KestrelHttpServer\test\Microsoft.AspNetCore.Server.Kestrel.Performance\BDN.Generated\binaries

// *** Build ***
C:\Program Files\dotnet\dotnet.exe compile-csc @D:\src\aspnet\KestrelHttpServer\test\Microsoft.AspNetCore.Server.Kestrel.Performance\obj\Release\netcoreapp1.1\dotnet-compile.rsp returned Exit Code 1
D:\src\aspnet\KestrelHttpServer\test\Microsoft.AspNetCore.Server.Kestrel.Performance\error CS2012: Cannot open 'D:\src\aspnet\KestrelHttpServer\test\Microsoft.AspNetCore.Server.Kestrel.Performance\bin\Release\netcoreapp1.1\Microsoft.AspNetCore.Server.Kestrel.Performance.dll' for writing -- 'The process cannot access the file 'D:\src\aspnet\KestrelHttpServer\test\Microsoft.AspNetCore.Server.Kestrel.Performance\bin\Release\netcoreapp1.1\Microsoft.AspNetCore.Server.Kestrel.Performance.dll' because it is being used by another process.'
// Result = Failure
```

Any ideias? I've even reset my machine to make sure no process was hanging on to that file, but it didn't help. I don't have VS open so that's not it either.

halter73 commented:
Did you compare #1218 to dev using these benchmarks?

benaadams commented:
@cesarbs had that issue after killing it halfway through and re-running (maybe VS?)

Deleted the bin/Release directory and ran again at cmd and got these results

```
                   Method |        Mean |    StdDev |  Gen 0 | Allocated |
------------------------- |------------ |---------- |------- |---------- |
                    Write |  51.5520 ns | 0.7567 ns |      - |       0 B |
             WriteChunked |  52.2010 ns | 0.8147 ns |      - |       0 B |
               WriteAsync |  91.7909 ns | 0.4175 ns | 0.0142 |      63 B |
        WriteAsyncChunked |  89.3543 ns | 0.5601 ns | 0.0142 |      63 B |
        WriteAsyncAwaited | 134.9429 ns | 1.1241 ns | 0.0138 |      63 B |
 WriteAsyncAwaitedChunked | 133.7213 ns | 0.7650 ns | 0.0139 |      63 B |
               ProduceEnd |  47.2052 ns | 0.4014 ns | 0.0170 |      63 B |
        ProduceEndChunked |  47.2801 ns | 0.1889 ns | 0.0172 |      63 B |
```

cesarbs commented:
@benaadams No luck. What fixes it for me is adding a no-op benchmark that gets run first - that one takes the hit and the rest runs fine.

I'll try to repro in a separate app, might be a bug in BenchmarkDotNet.

cesarbs commented:
Initially I thought this was due to the switch to `netcoreapp1.1`, but it actually reproes even in `dev` with the existing `RequestParsing` benchmarks. And this is after doing a lot of cleanup:

- Deleted NuGet cache
- Cleaned git repo
- Restarted the machine
- Rebuilt in `dev`
- Ran the tests

This is crazy.

cesarbs commented:
Here's the issue:

`dotnet run -c Release` -> OK
`dotnet run -c release` -> Fails

cesarbs commented:
https://github.com/dotnet/BenchmarkDotNet/issues/322

cesarbs commented:
Posted measurements that prompted this change here: https://github.com/aspnet/KestrelHttpServer/pull/1218#issuecomment-266158979

benaadams commented:
> Here's the issue:

How oddly random 😀
